src/main/java/org/vanilladb/core/storage/buffer/Buffer.java

```diff
@@ -164,12 +166,9 @@ public LogSeqNum lastLsn(){
 	 * @return a block ID
 	 */
 	public BlockId block() {
-		internalLock.readLock().lock();
-		try {
-			return blk;
-		} finally {
-			internalLock.readLock().unlock();
-		}
+		// Optimization
+		// blk will be modified only if no txs pin this buffer 
+		return blk;

```

Comment:
Do you think we need a `swapLock` here?

Reply:
![image](https://user-images.githubusercontent.com/25839129/143671382-f497212e-6085-4799-ab50-110bbb8cf19d.png)

If we could verify these three highlight methods are safe to get blk. We might not need a swap lock.
In current code, they get block after unpin the buffer, which is really danger.

One solution is to swap two lines of code
![image](https://user-images.githubusercontent.com/25839129/143671484-07295aa4-7e33-4431-b42b-080cce924e23.png)

Reply:
Sure. This is definitely a solution, but I am worried that other contributors in the future may not know this assumption. Maybe we should add a check or use something else to prevent other code from calling `block()` without pinning the buffer.